include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -1324,6 +1324,292 @@ namespace Differentiation
+     * function when using ADOL-C number types. It is, however, suitable for use
+     * in both serial and MPI routines.
+     *
+     * @author Jean-Paul Pelteret, 2016, 2017, 2018
+     */
+    template <enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperResidualLinearisation
+      : public ADHelperCellLevelBase<ADNumberTypeCode, ScalarType>
+    {
+    public:
+      /**
+       * Type definition for the floating point number type that is used in,
+       * and results from, all computations.
+       */
+      using scalar_type =
+        typename ADHelperBase<ADNumberTypeCode, ScalarType>::scalar_type;
+
+      /**
+       * Type definition for the auto-differentiation number type that is used
+       * in all computations.
+       */
+      using ad_type =
+        typename ADHelperBase<ADNumberTypeCode, ScalarType>::ad_type;
+
+      /**
+       * @name Constructor / destructor
+       */
+      //@{
+
+      /**
+       * The constructor for the class.
+       *
+       * @param[in] n_independent_variables The number of independent variables
+       * that will be used in the definition of the functions that it is
+       * desired to compute the sensitivities of. In the computation of
+       * $\mathbf{r}(\mathbf{X})$, this will be the number of inputs
+       * $\mathbf{X}$, i.e. the dimension of the domain space.
+       * @param[in] n_dependent_variables The number of scalar functions to be
+       * defined that will have a sensitivity to the given independent
+       * variables. In the computation of $\mathbf{r}(\mathbf{X})$, this will
+       * be the number of outputs $\mathbf{r}$, i.e. the dimension of the
+       * image space.
+       */
+      ADHelperResidualLinearisation(const unsigned int n_independent_variables,
+                                    const unsigned int n_dependent_variables);
+
+      /**
+       * Destructor
+       */
+      virtual ~ADHelperResidualLinearisation() = default;
+
+      //@}
+
+      /**
+       * @name Dependent variables
+       */
+      //@{
+
+      /**
+       * Register the definition of the cell residual vector
+       * $\mathbf{r}(\mathbf{X})$.
+       *
+       * @param[in] residual A vector of recorded functions that defines the
+       * residual. The components of this vector represents the dependent
+       * variables.
+       *
+       * @note For this class that expects only vector field of dependent
+       * variables, this function must only be called once per tape.
+       *
+       * @note For taped AD numbers, this operation is only valid in recording mode.
+       */
+      void
+      register_residual_vector(const std::vector<ad_type> &residual);
+
+      /**
+       * Evaluation of the residual for a chosen set of degree-of-freedom
+       * values. This corresponds to the computation the residual vector, i.e.
+       * @f[
+       *   \mathbf{r}(\mathbf{X}) \vert_{\mathbf{X}}
+       * @f]
```

Comment:
I think I would just inline this formula -- it looks large, but when printed it's actually going to look pretty small.

Reply:
In order to maintain consistency with the other classes' documentation, I think I'd prefer to leave this as is for now. I do my best to fix up all of the documentation (both merged, current and upcoming) as we're going along but this is just too much. If you feel really strongly about it then maybe we can look at consistently changing their formatting at the end.

Reply:
No, I don't feel really strongly :-) Rarely do, actually :-))